File: methods/matlab/RANDOMFOREST.m
% @file RANDOMFOREST.m
%
% Random Forest with matlab.
% Requires Statistics and Machine Learning toolbox installed

function randomforest(cmd)
% This program trains the Random Forest on the given labeled
% training set and then uses the trained classifier to classify the points
% in the given test set. Labels are expected to be the last row of the
% training set.
%
% Required options:
%     (-T) [string]    A file containing the test set.
%     (-t) [string]    A file containing the training set.
%     (-m) [int]       Min Leaf Size.
%     (-n) [int]       Number of Trees.

trainFile = regexp(cmd, '.*?-t ([^\s]+)', 'tokens', 'once');
testFile = regexp(cmd, '.*?-T ([^\s]+)', 'tokens', 'once');
minLeafSize = regexp(cmd, '.*?-m ([^\s]+)', 'tokens', 'once');
numTrees = regexp(cmd, '.*?-n ([^\s]+)', 'tokens', 'once');

% Load input dataset.
TrainData = csvread(trainFile{:});
TestData = csvread(testFile{:});

% Use the last row of the training data as the labels.
labels = TrainData(:,end);
% Remove the label row.
TrainData = TrainData(:,1:end-1);

% Create and train the classifier.
total_time = tic;
classifier = TreeBagger(str2num(numTrees{1}), TrainData, labels,... 
'MinLeafSize', str2num(minLeafSize{1}), 'Method', 'classification');
% Run Random Forest on the test dataset.
labels = predict(classifier, TestData);
disp(sprintf('[INFO ]   total_time: %fs', toc(total_time)))
% Save prediction of each class for test data.
csvwrite('predictions.csv', labels);

end

File: methods/weka/src/RANDOMFOREST.java
/**
 * @file RandomForest.java
 *
 * Random Forest with weka.
 */

import weka.classifiers.Classifier;
import weka.classifiers.RandomizableClassifier;
import weka.classifiers.trees.RandomForest;
import weka.core.Instances;
import weka.core.Utils;
import weka.core.converters.ConverterUtils.DataSource;
import weka.filters.Filter;
import weka.filters.unsupervised.attribute.NumericToNominal;
import java.io.File;
import java.io.FileWriter;
import java.io.BufferedWriter;
import weka.core.Attribute;
import java.util.List;
import java.util.ArrayList;
/**
 * This class use the weka libary to implement Random Forest.
 */
public class RANDOMFOREST {

  private static final String USAGE = String
      .format("This program trains the Decision Tree classifier on the given\n"
      + "labeled training set and then uses the trained classifier to classify\n"
      + "the points in the given test set.\n\n"
      + "Required options:\n"
      + "-T [string]     A file containing the test set.\n"
      + "-t [string]     A file containing the training set.\n"
      + "-M [int]        Minimum Instances per leaf");

  public static void main(String args[]) {
  Timers timer = new Timers();
    try {
      // Get the data set path.
      String trainFile = Utils.getOption('t', args);
      String testFile = Utils.getOption('T', args);
      String minLeafSize = Utils.getOption('M', args);

      if (trainFile.length() == 0 || testFile.length() == 0)
        throw new IllegalArgumentException();

      // Load train and test dataset.
      DataSource source = new DataSource(trainFile);
      Instances trainData = source.getDataSet();

      // Use the last row of the training data as the labels.
      trainData.setClassIndex((trainData.numAttributes() - 1));
      DataSource testsource = new DataSource(testFile);
      Instances testData = testsource.getDataSet();

      // Add pseudo class to the test set if no class information is provided.
      if (testData.numAttributes() < trainData.numAttributes()) {
        List<String> labelslist = new ArrayList<String>();
        for (int i = 0; i < trainData.classAttribute().numValues(); i++) {
          labelslist.add(trainData.classAttribute().value(i));
        }

        testData.insertAttributeAt(new Attribute("class", labelslist),
            testData.numAttributes());
      }

      // Use the last row of the training data as the labels.
      testData.setClassIndex((testData.numAttributes() - 1));

      timer.StartTimer("total_time");
      // Create and train the classifier.
      RandomForest cModel = new RandomForest();
      cModel.setOptions(weka.core.Utils.splitOptions("-M " + minLeafSize));
      cModel.buildClassifier(trainData);

      // Run Random Forest on the test dataset.
      // Write predicted class values for each intance to
      // benchmarks/weka_predicted.csv.
      double prediction = 0;
      try{
        File predictedlabels = new File("weka_predicted.csv");
        if(!predictedlabels.exists()) {
          predictedlabels.createNewFile();
        }
        FileWriter writer = new FileWriter(predictedlabels.getName(), false);

        for (int i = 0; i < testData.numInstances(); i++) {
          prediction = cModel.classifyInstance(trainData.instance(i));
          String pred = Double.toString(prediction+1);
          writer.write(pred);
          writer.write("\n");
        }

        writer.close();
      } catch(Exception e) {
        e.printStackTrace();
      }
      timer.StopTimer("total_time");
      timer.PrintTimer("total_time");

    } catch (IllegalArgumentException e) {
          System.err.println(USAGE);
      } catch (Exception e) {
        e.printStackTrace();
      }
  }
}
